pkg/porter/explain.go

```diff
@@ -0,0 +1,291 @@
+}
+
+type PrintableOutput struct {
+	Type        interface{} `json:"type" yaml:"type"`
+	ApplyTo     string      `json:"applyTo" yaml:"applyTo"`
+	Description string      `json:"description" yaml:"description"`
+}
+
+type PrintableParameter struct {
+	Type        interface{} `json:"type" yaml:"type"`
+	Default     interface{} `json:"default" yaml:"default"`
+	ApplyTo     string      `json:"applyTo" yaml:"applyTo"`
+	Description string      `json:"description" yaml:"description"`
+	Required    bool        `json:"required" yaml:"required"`
+}
+
+type PrintableAction struct {
+	Modifies bool `json:"modifies" yaml:"modifies"`
+	// Stateless indicates that the action is purely informational, that credentials are not required, and that the runtime should not keep track of its invocation
+	Stateless bool `json:"stateless" yaml:"stateless"`
+	// Description describes the action as a user-readable string
+	Description string `json:"description" yaml:"description"`
+}
+
+func (o *ExplainOpts) Validate(args []string, cxt *context.Context) error {
+	err := o.sharedOptions.Validate(args, cxt)
+	if err != nil {
+		return err
+	}
+	err = o.ParseFormat()
+	if err != nil {
+		return err
+	}
+	if o.Tag != "" {
+		o.File = ""
+		o.CNABFile = ""
+
+		return o.validateTag()
+	}
+	return nil
+}
+
+func (p *Porter) Explain(o ExplainOpts) error {
+	err := p.prepullBundleByTag(&o.BundleLifecycleOpts)
+	if err != nil {
+		return errors.Wrap(err, "unable to pull bundle before invoking credentials generate")
+	}
+
+	err = p.applyDefaultOptions(&o.sharedOptions)
+	if err != nil {
+		return err
+	}
+	err = p.ensureLocalBundleIsUpToDate(o.bundleFileOptions)
+	if err != nil {
+		return err
+	}
+	bundle, err := p.CNAB.LoadBundle(o.CNABFile, o.Insecure)
+	// Print Bundle Details
+
+	pb, err := generatePrintable(bundle)
+	if err != nil {
+		return errors.Wrap(err, "unable to print bundle")
+	}
+	switch o.Format {
+	case printer.FormatJson:
+		return printer.PrintJson(p.Out, pb)
+	case printer.FormatYaml:
+		return printer.PrintYaml(p.Out, pb)
+	case printer.FormatTable:
+		return p.printTable(pb)
+	default:
+		return fmt.Errorf("invalid format: %s", o.Format)
+	}
+}
+
+func generatePrintable(bun *bundle.Bundle) (*PrintableBundle, error) {
+	if bun == nil {
+		return nil, fmt.Errorf("expected a bundle")
+	}
+	pb := PrintableBundle{
+		Name:        bun.Name,
+		Description: bun.Description,
+		Version:     bun.Version,
+	}
+
+	actions := map[string]PrintableAction{}
+	for a, v := range bun.Actions {
+		pa := PrintableAction{}
+		pa.Description = v.Description
+		pa.Modifies = v.Modifies
+		pa.Stateless = v.Stateless
+		actions[a] = pa
+	}
+
+	creds := map[string]PrintableCredential{}
+	for c, v := range bun.Credentials {
+		pc := PrintableCredential{}
+		pc.Description = v.Description
+		pc.Required = v.Required
+
+		creds[c] = pc
+	}
+
+	params := map[string]PrintableParameter{}
+	for p, v := range bun.Parameters {
+		def, ok := bun.Definitions[v.Definition]
```

Comment:
Idea that this reminded me of whilst writin' similar code in Porter: We should add a helper func in cnab-go, `GetDefinition(paramOrOutput)` or some such (may need a function for each type), that'd return a parameter's/output's definition or return the corresponding error (not found/empty/etc.)

Reply:
That’s a good idea. Do you want to file an issue for it?

Reply:
Yuuuup https://github.com/deislabs/cnab-go/issues/134

Reply:
Thanks @vdice!